notebooks/logistic-regression.ipynb

```diff
@@ -647,6 +647,108 @@
     "else:\n",
     "    print(f\"logLoss: {evaluator.evaluate(cpu_transformed_df)}\")"
    ]
+  },
+  {
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "## Sparse Vectors"
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": null,
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    "def sparse_vectors_compat(EstimatorClass):\n",
+    "    from pyspark.ml.evaluation import MulticlassClassificationEvaluator\n",
+    "    from pyspark.ml.feature import CountVectorizer, RegexTokenizer\n",
+    "    from pyspark.sql import Row\n",
+    "    from sklearn.datasets import fetch_20newsgroups\n",
+    "\n",
+    "    # prepare dataframe\n",
+    "    try:\n",
+    "        twenty_train = fetch_20newsgroups(subset=\"train\", shuffle=True, random_state=42)\n",
```

Comment:
Does this take a while?   Done twice currently.

Reply:
Revised to move the dataframe preparation code out so that it is executed once.